### ISSUE TITLE
unreasonable vram cache used by falcon-40b

### ISSUE THREAD

**Issue by stupiding** [huggingface/text-generation-inference#532]
### System Info

Target: x86_64-unknown-linux-gnu
Cargo version: 1.69.0
Commit sha: 53aa9194c8c070afd19fa4660305dab2b280adf3
Docker label: sha-53aa919
nvidia-smi:
        NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1 
        3090 x 8 

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

```
import json
import torch
import asyncio
from text_generation import AsyncClient

async def main():
    client = AsyncClient("http://localhost:8082")

    free_mem = torch.cuda.mem_get_info('cuda:7')[0]
    print(f'free vram : {free_mem / 2 ** 30:.02f} GiBs')

    prompt = 'test '
    for length in [500, 1000, 1500]:
        request = client.generate_stream(
                prompt * length,
                do_sample=False,
                max_new_tokens=128,
                temperature=0.2,
                repetition_penalty=2.0,
                )

        text = ""
        async for response in request:
            if not response.token.special:
                text += response.token.text
        free_mem = torch.cuda.mem_get_info('cuda:7')[0]
        print(f'{i} token free vram: {free_mem / 2 ** 30:.02f} GiBs')

if __name__ == '__main__':
    asyncio.run(main())
~
```

1. launch the falcon-40b model `text-generation-launcher --model-id tiiuae/falcon-40b/ --sharded true --num-shard 8 --max-input-length 2048 --max-total-tokens 2560`
2. run the above test code `python local_test.py`, and the outputs like below
```
free vram : 12.67 GiBs
500 token free vram : 11.65 GiBs
1000 token free vram : 11.14 GiBs
1500 token free vram : 9.59 GiBs
```

### Expected behavior

1. from the 1500 token output, we can conclude that every token consumes about 1MB on each gpu, which is much larger than expected: 60 x 1 x 2 x 64 x 2B = 15KB
2. for the 1500 token request, the cache allocated for 500 and 1000 seems not reachable anymore, but are never released.

**Comment by OlivierDehaene** [comment]
Can you try on 0.9?

**Comment by stupiding** [comment]
> In any case, 0.9 doesn't have the same layout at all so I'm closing this for now.

I have tried with 0.9 docker:
Target: x86_64-unknown-linux-gnu
Cargo version: 1.70.0
Commit sha: 1da07e85aae8ce417dda3effd516691394dc31a1
Docker label: sha-1da07e8

If I haven't make it wrong, but this new docker consumes more vram at startup 
![image](https://github.com/huggingface/text-generation-inference/assets/12253817/69cc0ee0-7088-4fee-b1e6-8ba740af0c12)

And again, when I run the test code, with a little change from `for length in [500, 1000, 1500]:` to `for length in [500, 1000, 1500, 2000]:`, the vram continues to grow, after the 2000 token test, the nvidia-smi shows like below, which is nearly same as get_mem_info():
![image](https://github.com/huggingface/text-generation-inference/assets/12253817/0e6cb006-7e0a-4f01-9dfd-f242787c36e3)

![image](https://github.com/huggingface/text-generation-inference/assets/12253817/0c92d500-ed48-4df9-a051-b4498e1d99c9)
